node/network/statement-distribution/src/vstaging/candidates.rs

```diff
@@ -588,12 +593,682 @@ impl ConfirmedCandidate {
+			relay_hash,
+			group_index,
+			Some((candidate_head_data_hash_a, 1.into())),
+		);
+		assert_eq!(
+			candidates.by_parent,
+			HashMap::from([
+				((relay_hash, 1.into()), HashSet::from([candidate_hash_a])),
+				(
+					(candidate_head_data_hash_a, 1.into()),
+					HashSet::from([candidate_hash_b, candidate_hash_c])
+				)
+			])
+		);
+
+		// Advertise D with parent A.
+		candidates.insert_unconfirmed(
+			peer,
+			candidate_hash_d,
+			relay_hash,
+			group_index,
+			Some((candidate_head_data_hash_a, 1.into())),
+		);
+		assert_eq!(
+			candidates.by_parent,
+			HashMap::from([
+				((relay_hash, 1.into()), HashSet::from([candidate_hash_a])),
+				(
+					(candidate_head_data_hash_a, 1.into()),
+					HashSet::from([candidate_hash_b, candidate_hash_c, candidate_hash_d])
+				)
+			])
+		);
+
+		// Insert confirmed candidates and check parent hash index.
+
+		// Confirmation matches advertisement. Index should be unchanged.
+		candidates.confirm_candidate(candidate_hash_a, candidate_a, pvd_a, group_index);
+		assert_eq!(
+			candidates.by_parent,
+			HashMap::from([
+				((relay_hash, 1.into()), HashSet::from([candidate_hash_a])),
+				(
+					(candidate_head_data_hash_a, 1.into()),
+					HashSet::from([candidate_hash_b, candidate_hash_c, candidate_hash_d])
+				)
+			])
+		);
+		candidates.confirm_candidate(candidate_hash_b, candidate_b, pvd_b, group_index);
+		assert_eq!(
+			candidates.by_parent,
+			HashMap::from([
+				((relay_hash, 1.into()), HashSet::from([candidate_hash_a])),
+				(
+					(candidate_head_data_hash_a, 1.into()),
+					HashSet::from([candidate_hash_b, candidate_hash_c, candidate_hash_d])
+				)
+			])
+		);
+
+		// Confirmation does not match advertisement. Index should be updated.
+		candidates.confirm_candidate(candidate_hash_d, candidate_d, pvd_d, group_index);
+		assert_eq!(
+			candidates.by_parent,
+			HashMap::from([
+				((relay_hash, 1.into()), HashSet::from([candidate_hash_a])),
+				(
+					(candidate_head_data_hash_a, 1.into()),
+					HashSet::from([candidate_hash_b, candidate_hash_c])
+				),
+				((candidate_head_data_hash_c, 1.into()), HashSet::from([candidate_hash_d]))
+			])
+		);
+
+		// Make a new candidate for C with a different para ID.
+		let (new_candidate_c, new_pvd_c) = make_candidate(
+			relay_hash,
+			1,
+			2.into(),
+			candidate_head_data_b,
+			candidate_head_data_c.clone(),
+			Hash::from_low_u64_be(3000).into(),
+		);
+		candidates.confirm_candidate(candidate_hash_c, new_candidate_c, new_pvd_c, group_index);
+		assert_eq!(
+			candidates.by_parent,
+			HashMap::from([
+				((relay_hash, 1.into()), HashSet::from([candidate_hash_a])),
+				((candidate_head_data_hash_a, 1.into()), HashSet::from([candidate_hash_b])),
+				((candidate_head_data_hash_b, 2.into()), HashSet::from([candidate_hash_c])),
+				((candidate_head_data_hash_c, 1.into()), HashSet::from([candidate_hash_d]))
+			])
+		);
+	}
+
+	#[test]
+	fn pruning_unconfirmed_claims_maintains_parent_hash_index() {
+		todo!()
+	}
```

Comment:
I think pruning is already tested in `confirming_maintains_parent_hash_index`, but wanted to double-check before removing this.

Reply:
Looks like it to me